Add Google and Microsoft login providers

vvillena opened the pull request:
#24

victorvillenaosorio committed:
sessionController with endpoint for login and callback

victorvillenaosorio committed:
external login provider configuration

victorvillenaosorio committed:
appsettings configuration for external providers

victorvillenaosorio committed:
adding two needed libraries for registering google and microsoft login providers

victorvillenaosorio committed:
Merge from Version3, adding some changes

victorvillenaosorio committed:
remove from appsettings

victorvillenaosorio committed:
add missing changes

victorvillenaosorio committed:
Remove unused class
Add external settings to authorization settings
Add logic for registering external authorization to ConfigurationHelper and AuthorizationSettingsBuilder
Add routing and endpoints to startup, and changed registerExternalAuthentication call

victorvillenaosorio committed:
remove test controller

xaberue commented:
Hi @vvillena , there are some conflicts pending to solve

xaberue commented on ChustaSoft.Tools.Authorization.AspNet/Controllers/ExternalAuthorizationController.cs:
```diff
@@ -0,0 +1,58 @@
+{
+    [Controller]
+    [Route("web/[controller]")]
+    public class ExternalAuthorizationController : Controller
```
I think we can integrate this endpoints in the existing AuthorizationController, no need to add another one unless you are doing a separate manual controller registration in the DI (Not as I can see)

**Currently**, 
it is something like externalauthorization/login/{provider}
Additionally we have auth/[...] in the other Controller

**Suggested**
Integrate in the other controller, auth/external-login/{provider}

xaberue commented on ChustaSoft.Tools.Authorization.AspNet/Controllers/ExternalAuthorizationController.cs:
```diff
@@ -0,0 +1,58 @@
+            this.signInManager = signInManager;
+        }
+
+        [AllowAnonymous]
```
Remove testing method

xaberue commented on ChustaSoft.Tools.Authorization.AspNet/Controllers/ExternalAuthorizationController.cs:
```diff
@@ -0,0 +1,58 @@
+        [HttpGet("login/{redirectUrl}/callback", Name = "web-session-external-login-callback")]
+        public IActionResult ExternalLoginCallback([FromRoute]string redirectUrl)
+        {
+            //Example of how to retrieve login information:
```
Remove comments

xaberue commented on ChustaSoft.Tools.Authorization.TestBasic.WebAPI/Startup.cs:
```diff
@@ -44,7 +46,9 @@ public Startup(IConfiguration configuration)
-                .WithSqlServerProvider(BuildConnectionString());
+                .WithSqlServerProvider(BuildConnectionString());            
+
+            services.RegisterExternalAuthentication(opt => opt.SetExternalAuthorization(_configuration));
```
Not required, if is using appSettings file, has to take the configuration automatically from it

xaberue commented on ChustaSoft.Tools.Authorization/Configuration/AuthorizationSettings.cs:
```diff
@@ -11,6 +13,8 @@ public class AuthorizationSettings
         public bool StrongSecurityPassword { get; set; }
         public string DefaultCulture { get; set; }
 
+        public List<(ExternalAuthenticationProviders Provider, (string ClientId, string ClientSecret))> ExternalSettings { get; set; }
```
It is possible to chage as an IEnumerable instead of a concrete List implementation?
Better if possible, could be changed to a IDictionary ?? should not be more than one provider configured...

Additionally, is missing the the anoymous first tuple  object name

Final suggestion then:
 public **IDictionary**<ExternalAuthenticationProviders **Name**, (string ClientId, string ClientSecret) **Config**> **ExternalProviders** { get; set; }

vvillena replied:
I thought we talked about having it as a list of tuples, maybe I understood it wrong. I've created an object to map the properties. And changed the list to an IEnumerable.

xaberue replied:
**IDictionary<string, TupleObject> ExternalProviders { get; set; }**


Sorry bro <3

vvillena replied:
Changed by IDictionary

xaberue commented on ChustaSoft.Tools.Authorization/Configuration/AuthorizationSettingsBuilder.cs:
```diff
@@ -1,9 +1,18 @@
 
 namespace ChustaSoft.Tools.Authorization
 {
+    public enum ExternalAuthenticationProviders
```
Move to another file

xaberue commented on ChustaSoft.Tools.Authorization/Configuration/AuthorizationSettingsBuilder.cs:
```diff
@@ -56,11 +65,28 @@ public AuthorizationSettingsBuilder SetMinutesToExpire(int minutesToExpire)
-        public AuthorizationSettings Build() => _authorizationSettings;
+        public AuthorizationSettingsBuilder SetExternalAuthorization(IConfiguration configuration)
+        {
+            var externalProviders = Enum.GetValues(typeof(ExternalAuthenticationProviders));
```
CommonNET EnumsHelper could help with that, being Type safe: **EnumsHelper.GetEnumList<ExternalAuthenticationProviders>();**

xaberue commented on ChustaSoft.Tools.Authorization/Configuration/AuthorizationSettingsBuilder.cs:
```diff
@@ -56,11 +65,28 @@ public AuthorizationSettingsBuilder SetMinutesToExpire(int minutesToExpire)
-    }
+            foreach (ExternalAuthenticationProviders externalProvider in externalProviders)
+            {
+                string externalProviderName = Enum.GetName(typeof(ExternalAuthenticationProviders), externalProvider);
```
Same than above: EnumsHelper .GetDescription(externalProvider);

xaberue commented on ChustaSoft.Tools.Authorization/Configuration/AuthorizationSettingsBuilder.cs:
```diff
@@ -56,11 +65,28 @@ public AuthorizationSettingsBuilder SetMinutesToExpire(int minutesToExpire)
         }
 
-        public AuthorizationSettings Build() => _authorizationSettings;
+        public AuthorizationSettingsBuilder SetExternalAuthorization(IConfiguration configuration)
```
IoC change:
Credentials from IConfiguration should be moved to ConfigurationHelper, we are adding the responsibility there to know about IConfiguration to the AuthorizationSettingsBuilder  when it is not required. Additionally, we are not allowing the tool to configure it from another mechanism different than appSettings.

The builder is thought to be prepared for configuring evth without Configuration

vvillena replied:
My fault, I've moved the logic and now I think it matches the responsability of each class.

xaberue commented on ChustaSoft.Tools.Authorization/Configuration/AuthorizationSettingsBuilder.cs:
```diff
@@ -69,6 +95,8 @@ public interface IAuthorizationSettingsBuilder : IBuilder<AuthorizationSettings>
         AuthorizationSettingsBuilder SetMinutesToExpire(int minutesToExpire);
         AuthorizationSettingsBuilder SetPasswordSecurity(bool flag, int minLength);
         AuthorizationSettingsBuilder SetSiteName(string siteName);
+        AuthorizationSettingsBuilder SetExternalAuthorization(IConfiguration configuration);
```
Same IoC change than above

xaberue commented on ChustaSoft.Tools.Authorization/Helpers/SecurityKeyHelper.cs:
```diff
@@ -8,10 +8,13 @@ public static class SecurityKeyHelper
         {
-            var secretKeyBytes = Encoding.UTF8.GetBytes(privateKey);
-            var signingKey = new SymmetricSecurityKey(secretKeyBytes);
+            if (!string.IsNullOrEmpty(privateKey)) 
```
Remove this validation. 
If private key is not setted up, should throw an Exception because it is mandatory for the right behaviour of the tool

xaberue commented on ChustaSoft.Tools.Authorization/Configuration/ConfigurationHelper.cs:
```diff
@@ -68,6 +100,11 @@ public static IdentityBuilder WithCustomUserAction<TCustomUserAction>(this Ident
         #region Private methods
+        private static bool SettingsHaveValues((ExternalAuthenticationProviders Provider, (string ClientId, string ClientSecret)) externalAuthenticationSetting)
+        {
+            return !string.IsNullOrEmpty(externalAuthenticationSetting.Item2.ClientId)
```
If tuples params has custom names you can override Item1, Item2, ... 
Semantic sugar :)

victorvillenaosorio committed:
moving business logic
joining controllers
getting rid of tuples list, by an ienumerable of objects
little refactor

vvillena committed:
Merge branch 'version3' into #24

victorvillenaosorio committed:
Fix change IEnumerable by IDictionary of providers
Logic change to make it work with IDictionary change

victorvillenaosorio committed:
Merge branch '#24' of https://github.com/ChustaSoft/Authorization into #24

xaberue commented on ChustaSoft.Tools.Authorization.AspNet/Controllers/AuthorizationController.cs:
```diff
@@ -17,16 +19,21 @@ public class AuthorizationController : ApiControllerBase<AuthorizationController
         #region Fields
 
         private readonly ISessionService _sessionService;
+        private readonly SignInManager<User> signInManager;
```
We can't directly use SignInManager here, ISessionService encapsulates Services that at same time are using Managers,

In that case should be at IUserService, and then by using it throug ISessionService

With this approach we can have a problem, we are forcing the implementation to use User instead of generic types

xaberue commented on ChustaSoft.Tools.Authorization.TestBasic.WebAPI/Startup.cs:
```diff
@@ -44,7 +45,9 @@ public Startup(IConfiguration configuration)
-                .WithSqlServerProvider(BuildConnectionString());
+                .WithSqlServerProvider(BuildConnectionString());            
+
+            services.RegisterExternalAuthentication(_configuration);
```
This should be performed automatically through the settings or through the builder.

Please check it

vvillena replied:
Logic for external authentication registering has been extracted, and method has been overloaded to allow accept IConfiguration settings or authorization settings builder

xaberue replied:
Thanks for have in mind! This was required.

However, one point is still missing, this line should not be required, instead of using this additional line in the configuration, shoud be performed automatically (Private method instead of public one) doing it just in case of having external configurations added to the Settings DTO

xaberue commented on ChustaSoft.Tools.Authorization.TestBasic.WebAPI/appsettings.json:
```diff
@@ -1,6 +1,6 @@
 {
   "ConnectionStrings": {
-    "AuthorizationConnection": ""
+    "AuthorizationConnection": "Server=myServerName\nyInstanceName;Database=nyDataBase;User Id=myUsername;Password=myPassword;"
```
Remove it before PR integration, better leave the user to directly paste everything instead of override settings one by one

xaberue commented on ChustaSoft.Tools.Authorization/Configuration/ConfigurationHelper.cs:
```diff
@@ -34,11 +35,41 @@ public static IdentityBuilder RegisterAuthorization(this IServiceCollection serv
         }
 
-        public static IdentityBuilder RegisterAuthorization(this IServiceCollection services, string privateKey, Action<IAuthorizationSettingsBuilder> settingsBuildingAction) 
+        public static void RegisterExternalAuthentication(this IServiceCollection services, IConfiguration configuration)
```
As said before, this should be automatically registered once the settings are loaded, regardless of whether it is through the appSettings file or the builder.

vvillena replied:
Logic for external authentication registering has been extracted, and method has been overloaded to allow accept IConfiguration settings or authorization settings builder

xaberue replied:
It is quite close to be as suggested, now this logic should be moved to the main method (RegisterAuthorization) properly loading the settings DTO and doing this logic in a private method.

The idea for this is to have a single line configuration in the Startup. Multiple configuration lines could create confussion and errors for future implementations. Single line eases the set up.

xaberue reviewed (changes requested):
Close to be finished, but still requires some important changes to fully meet current behavior, and expected usage and settings.

Thanks a lot in advance!

victorvillenaosorio committed:
Fix PR:
- Remove signinManager dependency from AuthorizationController and move logic to IUserservice
- Extract logic for register external authentication, overload method to allow registering it by settings or builder
- Remove testing connection string from appsettings.json

xaberue commented on ChustaSoft.Tools.Authorization.TestBasic.WebAPI/Startup.cs:
```diff
@@ -71,6 +74,16 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env, IService
+
+            app.UseEndpoints(endpoints =>
+            {
+                endpoints.MapControllerRoute(
```
I think we only have to map the Auth controller, however, we are forcing the client project to use this specific mapping

vvillena replied:
Code removed, as we are adding the AuthorizationController from the Chustasoft.ToolsAuthorization.AspNet assembly in both test projects with the .AddAuthorizationControllers() method

xaberue commented on ChustaSoft.Tools.Authorization.TestCustom.WebAPI/Startup.cs:
```diff
@@ -52,13 +52,21 @@ public void ConfigureServices(IServiceCollection services)
                 .WithUserCreatedAction<CustomUser, CustomUserAction>()
                 .WithSqlServerProvider<AuthCustomContext, CustomUser, CustomRole>(BuildConnectionString());
 
+            services.RegisterExternalAuthentication(_configuration);
```
Same explained in the Basic example

vvillena replied:
We are adding the AuthorizationController from the Chustasoft.ToolsAuthorization.AspNet assembly in both test projects with the .AddAuthorizationControllers() method

xaberue commented on ChustaSoft.Tools.Authorization/Configuration/AuthorizationSettingsBuilder.cs:
```diff
@@ -56,11 +60,16 @@ public AuthorizationSettingsBuilder SetMinutesToExpire(int minutesToExpire)
         }
 
-        public AuthorizationSettings Build() => _authorizationSettings;
+        public AuthorizationSettingsBuilder AddExternalAuthorization(ExternalAuthenticationProviders providerName, ExternalAuthenticationSettings externalAuthenticationSettings)
```
Suggestions here:

This is perfectly valid, but we can add another configuration option, Keep this as it is, and add a new AddExternalAuthorization overloaded method with the following parameters:

- providerName (ExternalAuthenticationProviders)
- clientId (string)
- secretId (string)

This alllow the user not only to send a new ExternalAuthenticationSettings, also can ease configuration sending parameters separated

vvillena replied:
Overloaded method added

xaberue reviewed (changes requested):
Sorry bro for the new suggestions and for reopening one, it is even closer to being ready, but we can still improve some details, there is only one important change required in relation to the way to configure it

victorvillenaosorio committed:
Fix PR: Overload method AddExternalAuthorization, Remove route mapping as we are registering the AuthorizationController from the assembly of Chustasoft.ToolsAuthorization.AspNet

victorvillenaosorio committed:
missing on interface

xaberue committed:
PR final changes

xaberue committed:
Code final cleanup, syntax normalizing and NuGet updates

xaberue committed:
Removed unnecessary method on ConfigurationHelper
